### PR TITLE
[CMake] Require cmake >= 3.23 but only on capnp builds

### PR DESCRIPTION
**Pull request by teqdruid**
"Fixes" #5242.

### CODE REVIEW

**Review by mikeurbach — COMMENTED** [review]
The note in the docs says to call this at the beginning of the top-level, even before `project`: https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html. I think the point is you need to declare the minimum version early, before you start using other bits of CMake. But I guess this works (I just tried it out to confirm, and it does give the error as expected).

It seems a little weird to have it this way though. This command is expected to be right at the top of the file, and the fact that there's also a check later in the file gated by some condition feels surprising. I'm not sure what CMake best practice is and if this is reasonable, but I'm curious what others think.

**Comment by darthscsi** [comment]
Debian stable has 3.18.4
Ubuntu 23.04 has 3.25.1
Ubuntu 22.10 has 3.24.2
Ubuntu 22.04LTS has 3.22.1
CentOS8 has 3.20.2

Which is probably to say 3.23 as default is probably pushing it too much, but not by much.  I'm not sure on homebrew or bsds.

**Comment by mikeurbach** [comment]
> An alternate might be to guard the cmake code based on version rather than changing the minimum.

That makes sense to me. So if the version is not 3.23 or greater, just disable capnp?

Also, @nandor already just bumped the minimum to 3.23 in https://github.com/llvm/circt/commit/f7e40324a67097acf9019c54e5bef30400a92da3, so we should probably revert that if we don't want to move that far ahead.

**Comment by teqdruid** [comment]
> > An alternate might be to guard the cmake code based on version rather than changing the minimum.
> 
> That makes sense to me. So if the version is not 3.23 or greater, just disable capnp?

If you ask for a capnp build, you should get either a capnp build or an error, so I don't like this. If you happen to have capnp installed and it errors out, you can always disable capnp.

So I vote to merge this PR as-is.

**Comment by mikeurbach** [comment]
> If you ask for a capnp build, you should get either a capnp build or an error

I think I probably misinterpreted the idea with my previous suggestion. Rather than disabling capnp, you can check the version and produce an error, like upstream MLIR does when Python is enabled: https://github.com/llvm/llvm-project/blob/25a4a9360e1dedb572dc007cf8cc97dfc3efdeac/mlir/cmake/modules/MLIRDetectPythonEnv.cmake#L5-L10

That way it's still enforcing the requirement, with a nice error, without having multiple calls to `cmake_minimum_version`. It's having multiple places call `cmake_minimum_version` that I'm concerned about with the current PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -413,6 +413,17 @@ option(CAPNP_DISABLE "Disable Cap'nProto (needed for cosimulation).")
 if (CAPNP_DISABLE)
   message (STATUS "Disabling Cap'nProto.")
 else()
+  # Some of our CMake code which is only used in Capnp builds uses newer CMake
+  # features. Since most compilations are not capnp builds, only require the new
+  # version for those builds.
+  if(CMAKE_VERSION VERSION_LESS "3.23.0")
+    message(SEND_ERROR
+        "Building CIRCT with Cap'nProto features enabled is CMake features "
```
Comment (mikeurbach) [inline CMakeLists.txt:421]: `is CMake features` seems off... `uses CMake features`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -413,6 +413,17 @@ option(CAPNP_DISABLE "Disable Cap'nProto (needed for cosimulation).")
...
+        "that require at least version 3.23. Recommend upgrading to 3.23+ "
+        "for full support. Detected current version: ${CMAKE_VERSION}.If you "
```
Comment (mikeurbach) [inline CMakeLists.txt:423]: nit: space after the period in `.If`

**Review by mikeurbach — APPROVED** [review]
Left a couple nitpicky comments on the error message, since error messages leave an impression. But in general I think this looks great, thanks for switching to doing it this way.

File: CMakeLists.txt (hunk 1 above)
Comment (teqdruid) [reply CMakeLists.txt]: I just copy-pasted from the python stuff and didn't copy edit. Oops!

**Pull request merged**